spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java

```diff
@@ -250,12 +288,59 @@ private List<ManifestFile> writeUnpartitionedManifests(
   private List<ManifestFile> writePartitionedManifests(
       ManifestContent content, Dataset<Row> manifestEntryDF, int numManifests) {
 
+    // Extract desired clustering criteria into a dedicated column
+    Dataset<Row> clusteredManifestEntryDF;
+
+    if (partitionClusteringFunction != null) {
+      LOG.info(
+          "Sorting manifests for specId {} using custom clustering function",
+          spec.specId(),
+          partitionClusteringFunction);
+      Types.StructType partitionType = DataFile.getType(table.spec().partitionType());
+      StructType dataFileSchema = manifestEntryDF.select("data_file.*").schema();
+
+      // Create a UDF to wrap the custom partitionClusteringFunction call
+      UserDefinedFunction clusteringUdf =
+          functions.udf(
+              new CustomDataFileClusteringUdf(
+                  this.partitionClusteringFunction, partitionType, dataFileSchema),
+              DataTypes.StringType);
+      // Apply supplied partitionSortFunction function to the data_file datums within this dataframe
+      // The results are stored as a String in the new __clustering_column__
+      clusteredManifestEntryDF =
+          manifestEntryDF.withColumn(
+              CUSTOM_CLUSTERING_COLUMN_NAME, clusteringUdf.apply(col("data_file")));
+    } else if (partitionFieldClustering != null) {
+      LOG.info(
+          "Clustering manifests for specId {} by partition columns by {} ",
+          spec.specId(),
+          partitionFieldClustering);
+
+      // Map the top level partition column names to the column name referenced within the manifest
+      // entry dataframe
+      Column[] actualPartitionColumns =
+          partitionFieldClustering.stream()
+              .map(p -> col("data_file.partition." + p))
+              .toArray(Column[]::new);
+
+      // Form a new temporary column to cluster manifests on, based on the custom clustering columns
+      // order provided
+      clusteredManifestEntryDF =
+          manifestEntryDF.withColumn(
+              CUSTOM_CLUSTERING_COLUMN_NAME, functions.struct(actualPartitionColumns));
+    } else {
+      clusteredManifestEntryDF =
```

Comment:
Rather than having a third branch we should probably just have `partitionFieldClustering` set to data_file.partition by default somehow. Just trying to see if we have a way to remove the complexity here.

Reply:
I'll give it a shot.